### PR TITLE
feat: option disabling parsing in the onRequest phase

### PR DESCRIPTION
**Pull request by vitaly-sazonov**
Hi!

Sometimes you need to receive cookies in the middleware, and then repeated parsing in the onRequest phase will be superfluous. The "addHookOnRequest: false" option can disable adding a hook, leaving all other functionality.

#### Checklist

- [*] run `npm run test` and `npm run benchmark`
- [*] tests and/or benchmarks are included
- [*] documentation is changed or added
- [*] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
why would you use this module if not for setting the cookies?

**Comment by vitaly-sazonov** [comment]
just to setting cookies. 

But the middleware phase is executed before the onRequest phase. If you need to work with cookies in the middleware phase, then you will have to parse the cookies in the middleware itself, and after that the plugin will parse the cookies again. extra work

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@fastify/cookie",
-  "version": "8.0.0",
+  "version": "8.0.1",
```
Comment (Uzlopak) [inline package.json:3]: revert this please, we increment on release and the one who releases sets the new version.

File: plugin.js [hunk 2]
```diff
@@ -63,6 +63,7 @@ function onReqHandlerWrapper (fastify) {
 
 function plugin (fastify, options, next) {
   const secret = options.secret
+  const addHookOnRequest = options.addHookOnRequest ?? true
```
Comment (Uzlopak) [inline plugin.js:66]: can we rename the flag? It describes what we do in plugin, but if you consume the plugin in your implementation, you are forced to check the sourcecode/readme of fastify/cookie to understand what it means. 

it should be something like parseCookieOnRequest or autoParseCookieOnRequest.

File: test/cookie.test.js [hunk 3]
```diff
@@ -847,3 +847,39 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+  fastify.inject({
+    method: 'GET',
+    url: '/disable',
+    headers: {
+      cookie: 'bar=bar'
+    }
+  }, (err, res) => {
+    t.error(err)
+    t.equal(res.statusCode, 200)
+    t.same(JSON.parse(res.body), { hello: 'world' })
```
Comment (Uzlopak) [inline test/cookie.test.js:883]: there is imho no need to check the body

File: test/cookie.test.js [hunk 4]
```diff
@@ -847,3 +847,39 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+  }
+
+  fastify.addHook('preParsing', (req, reply, payload, done) => {
+    t.notOk(req.cookies, null)
+    done()
+  })
+
+  fastify.get('/disable', (req, reply) => {
+    t.notOk(req.cookies, null)
+    reply.send({ hello: 'world' })
```
Comment (Uzlopak) [inline test/cookie.test.js:871]: just sending empty body should be enough if you remove the body assertioni in line 883

File: test/cookie.test.js [hunk 5]
```diff
@@ -847,3 +847,39 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+    })
+  }
+
+  fastify.addHook('preParsing', (req, reply, payload, done) => {
+    t.notOk(req.cookies, null)
+    done()
+  })
+
+  fastify.get('/disable', (req, reply) => {
+    t.notOk(req.cookies, null)
```
Comment (Uzlopak) [inline test/cookie.test.js:870]: what does this mean? I think the assertion is wrong. t.notOk is` t.notOk(obj, message, extra)` so null makes no sense for message

you want to do

t.equal(req.cookies, null) 
t.not(req.cookies, null)

File: test/cookie.test.js [hunk 6]
```diff
@@ -847,3 +847,39 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+
+  for (const hook of ['preValidation', 'preHandler']) {
+    fastify.addHook(hook, (req, reply, done) => {
+      t.notOk(req.cookies)
+      done()
+    })
+  }
+
+  fastify.addHook('preParsing', (req, reply, payload, done) => {
+    t.notOk(req.cookies, null)
```
Comment (Uzlopak) [inline test/cookie.test.js:865]: what does this mean? I think the assertion is wrong. t.notOk is` t.notOk(obj, message, extra)` so null makes no sense for message

you want to do

t.equal(req.cookies, null) 
t.not(req.cookies, null)

File: test/cookie.test.js [hunk 7]
```diff
@@ -847,3 +847,39 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+
+// addHookOnRequest: false
+test('disable parses incoming cookies', (t) => {
+  t.plan(7)
+  const fastify = Fastify()
+  fastify.register(plugin, { addHookOnRequest: false })
+
+  for (const hook of ['preValidation', 'preHandler']) {
+    fastify.addHook(hook, (req, reply, done) => {
+      t.notOk(req.cookies)
```
Comment (Uzlopak) [inline test/cookie.test.js:859]: what does this mean? can you make a more explicit check? like compare with undefined or null?

**Comment by Uzlopak** [comment]
I think this change is dangerous in regard #53 . So it could be that somebody deactivates cookie parsing on request but the cookie object is then not resetted. Imho we need also a flag resetCookieObjectOnRequest or a good example in the readme, to avoid that sec. vuln.

**Comment by Uzlopak** [comment]
Adding @climba03003 as reviewer for the potential sec. implications.

**Comment by climba03003** [comment]
> I think this change is dangerous in regard https://github.com/fastify/fastify-cookie/issues/53

No, the default should be `null` in here. It does not revert to the old behavior.


To the problem itself, I recommend to allow custom hook stage instead of a flag to disable the hook. It provides more flexibility to developer.

```js
fastify.register(FastifyCookie, { hook: 'onRequest' }) // add to onRequest hook
fastify.register(FastifyCookie, { hook: 'preHandler' }) // add to preHandler hook
fastify.register(FastifyCookie, { hook: false }) // disable parsing
```

File: plugin.js (hunk 2 above)
Comment (mcollina) [reply plugin.js:66]: I would prefer we add a `hook` option instead that default to `onRequest` but that it can also be `false`. If `false` we are not setting the hook at all.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: plugin.js (hunk 2 above)
Comment (vitaly-sazonov) [reply plugin.js:66]: yes, of course you can rename the flag. Let's think and choose some option:
 - parseCookieOnRequest
 - autoParseCookieOnRequest
 - autoParseCookie

File: test/cookie.test.js (hunk 3 above)
Comment (vitaly-sazonov) [reply test/cookie.test.js]: yes, thanks, fixed it

File: test/cookie.test.js (hunk 6 above)
Comment (vitaly-sazonov) [reply test/cookie.test.js]: yes, thanks, fixed it

File: plugin.js (hunk 2 above)
Comment (mcollina) [reply plugin.js:66]: see my comment

File: README.md [hunk 8]
```diff
@@ -33,8 +33,8 @@ const fastify = require('fastify')()
 
 fastify.register(require('@fastify/cookie'), {
   secret: "my-secret", // for cookies signature
-  addHookOnRequest: false, // disable cookie parsing on the onRequest phase
-  parseOptions: {}     // options for parsing cookies
+  hook: false, // disable cookie autoparsing or set autoparsing on any of the hooks: 'onRequest', 'preHandler', 'preParsing', 'preValidation', 'preSerialization'. default: onRequest
```
Comment (Uzlopak) [inline README.md:36]:
```suggestion
  hook: false, // disable cookie autoparsing or set autoparsing on any of the hooks: 'onRequest', 'preHandler', 'preParsing', 'preValidation', 'preSerialization'. default: 'onRequest'
```

File: plugin.js [hunk 9]
```diff
@@ -63,7 +63,7 @@ function onReqHandlerWrapper (fastify) {
 
 function plugin (fastify, options, next) {
   const secret = options.secret
-  const addHookOnRequest = options.addHookOnRequest ?? true
+  const hook = options.hook ?? 'onRequest'
```
Comment (Uzlopak) [inline plugin.js:66]: looks odd and could result in a error. I could set hook to true. and it result in an error. 

maybe write a little function which returns onRequest if it is true, false if false, string return the string.

File: README.md (hunk 8 above)
Comment (vitaly-sazonov) [reply README.md:36]: I think to exclude 'preSerialization' and 'PreRapsing', they have different arguments in the handler

Comment (mcollina) [reply README.md:36]: please set it `'onRequest'` here.

File: types/plugin.d.ts [hunk 10]
```diff
@@ -125,9 +125,17 @@ declare namespace fastifyCookie {
     signed?: boolean;
   }
 
+  enum HookType {
+    onRequest = "onRequest",
+    preHandler = "preHandler",
+    preParsing="preParsing",
+    preValidation="preValidation",
+    preSerialization="preSerialization"
```
Comment (mcollina) [inline types/plugin.d.ts:133]:
```suggestion
    preParsing = "preParsing",
    preValidation = "preValidation",
    preSerialization = "preSerialization"
```

**Review by mcollina — CHANGES_REQUESTED** [review]

File: plugin.js (hunk 9 above)
Comment (vitaly-sazonov) [reply plugin.js:66]: ok, may throw an error if the hook doesn't match 'preHandler', 'onRequest', 'preValidation' and false?

File: types/plugin.d.ts (hunk 10 above)
Comment (Uzlopak) [reply types/plugin.d.ts:133]: actually we dont need a enum here. 

type HookType = 'onRequest' | 'preHandler' | 'preParsing' | 'preValidation' | 'preSerialization';

File: plugin.js (hunk 9 above)
Comment (mcollina) [reply plugin.js:66]: Sure thing.

**Review by jsumners — COMMENTED** [review]
> But the middleware phase is executed before the onRequest phase. If you need to work with cookies in the middleware phase, then you will have to parse the cookies in the middleware itself, and after that the plugin will parse the cookies again before the route handler

What does that mean? Fastify does not have any such "phase" as Fastify does not use the "middleware" concept.

File: plugin.js [hunk 11]
```diff
@@ -61,8 +63,21 @@ function onReqHandlerWrapper (fastify) {
   }
 }
 
+function getHook (hook, next) {
+  const hooks = {
```
Comment (Uzlopak) [inline plugin.js:67]: nitpick: Would have used any array here.

Comment (vitaly-sazonov) [reply plugin.js:67]: yes, but accessing the object is a bit faster :)

File: test/cookie.test.js [hunk 12]
```diff
@@ -847,3 +847,59 @@ test('should not decorate fastify, request and reply if no secret was provided',
     message: 'req.unsignCookie is not a function'
   })
 })
+
+test('disable parses incoming cookies', (t) => {
```
Comment (Uzlopak) [inline test/cookie.test.js:851]:
```suggestion
test('dont add default @fastify/cookie hook-function to fastify if hook-option to false  ', (t) => {
```

File: test/cookie.test.js [hunk 13]
```diff
@@ -847,3 +847,59 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+    headers: {
+      cookie: 'bar=bar'
+    }
+  }, (err, res) => {
+    t.error(err)
+    t.equal(res.statusCode, 200)
+  })
+})
+
+test('error if option is invalid', (t) => {
```
Comment (Uzlopak) [inline test/cookie.test.js:885]:
```suggestion
test('result in an error if hook-option is set to an invaid value', (t) => {
```

File: test/cookie.test.js [hunk 14]
```diff
@@ -847,3 +847,59 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+  t.plan(1)
+  const fastify = Fastify()
+
+  fastify.register(plugin, { hook: true })
+
+  fastify.get('/error', (req, reply) => {
+    console.log('/error')
```
Comment (Uzlopak) [inline test/cookie.test.js:892]: no logs please 
```suggestion
```

File: test/cookie.test.js [hunk 15]
```diff
@@ -847,3 +847,59 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+    t.error(err)
+    t.equal(res.statusCode, 200)
+  })
+})
+
+test('error if option is invalid', (t) => {
+  t.plan(1)
+  const fastify = Fastify()
+
+  fastify.register(plugin, { hook: true })
```
Comment (Uzlopak) [inline test/cookie.test.js:889]: I would have expected to throw already here. strange

File: test/cookie.test.js (hunk 14 above)
Comment (vitaly-sazonov) [reply test/cookie.test.js:892]: yes, thanks, my mistake

File: test/cookie.test.js (hunk 15 above)
Comment (Uzlopak) [reply test/cookie.test.js:889]: Also I had a typo in my suggestion. Should be "invalid" and not "invaid"

The whole test is actually this:

```typescript
test('result in an error if hook-option is set to an invalid value', async (t) => {
  t.plan(1)
  const fastify = Fastify()

  t.rejects(() => fastify.register(plugin, { hook: true }), new Error("You can set false or use only 'onRequest' , 'preParsing' , 'preValidation' , 'preHandler' hooks"))
})

```

Comment (vitaly-sazonov) [reply test/cookie.test.js:889]: yes, it would indeed be better. corrected

File: plugin.js [hunk 16]
```diff
@@ -61,8 +63,21 @@ function onReqHandlerWrapper (fastify) {
...
+    onRequest: 'onRequest',
+    preParsing: 'preParsing',
+    preValidation: 'preValidation',
+    preHandler: 'preHandler',
+    [false]: false
+  }
+
+  return hooks[hook] ?? next(new Error('You can set false or use only \'onRequest\' , \'preParsing\' , \'preValidation\' , \'preHandler\' hooks'))
```
Comment (Uzlopak) [inline plugin.js:75]: I think rephrasing would help. prefixing it with @fastify/cookie helps to debug it faster. 

```suggestion
  return hooks[hook] ?? next(new Error('@fastify/cookie: Invalid value provided for the hook-option. You can set the hook-option only to false, \'onRequest\' , \'preParsing\' , \'preValidation\' or \'preHandler\''))
```

File: plugin.js [hunk 17]
```diff
@@ -51,7 +51,9 @@ function fastifyCookieClearCookie (reply, name, options) {
 }
 
 function onReqHandlerWrapper (fastify) {
-  return function fastifyCookieOnReqHandler (fastifyReq, fastifyRes, done) {
+  return function fastifyCookieHandler (...argv) {
+    const fastifyReq = argv[0]
+    const done = argv.pop()
```
Comment (Eomm) [inline plugin.js:56]: Why this change?

File: types/plugin.d.ts [hunk 18]
```diff
@@ -125,8 +125,11 @@ declare namespace fastifyCookie {
     signed?: boolean;
   }
 
+  type HookType = 'onRequest' | 'preParsing' | 'preValidation' | 'preHandler'  | 'preSerialization';
+
   export interface FastifyCookieOptions {
     secret?: string | string[] | Signer;
+    hook?: HookType| false;
```
Comment (Eomm) [inline types/plugin.d.ts:132]:
```suggestion
    hook?: HookType | false;
```

File: test/cookie.test.js [hunk 19]
```diff
@@ -847,3 +847,49 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+  fastify.register(plugin, { hook: false })
+
+  for (const hook of ['preValidation', 'preHandler']) {
+    fastify.addHook(hook, (req, reply, done) => {
+      t.equal(req.cookies, null)
+      done()
+    })
+  }
+
+  fastify.addHook('preParsing', (req, reply, payload, done) => {
```
Comment (Eomm) [inline test/cookie.test.js:863]: You should be able to add this hook to the hook array above by defining:

```js
async (req) => {
 t.equal(req.cookies, null)
}
```

File: test/cookie.test.js [hunk 20]
```diff
@@ -847,3 +847,49 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+})
+
+test('result in an error if hook-option is set to an invalid value', (t) => {
+  t.plan(1)
+  const fastify = Fastify()
+
+  t.rejects(
+    () => fastify.register(plugin, { hook: true }),
+    {},
+    { skip: true },
```
Comment (Eomm) [inline test/cookie.test.js:892]:
```suggestion
```

**Review by Eomm — CHANGES_REQUESTED** [review]

File: plugin.js (hunk 17 above)
Comment (vitaly-sazonov) [reply plugin.js:56]: on some hooks, the done function is the fourth argument

File: test/cookie.test.js (hunk 20 above)
Comment (Uzlopak) [reply test/cookie.test.js:892]: I dont even know this notation :D. Why not simply the error as second param?

File: plugin.js [hunk 21]
```diff
@@ -61,8 +63,21 @@ function onReqHandlerWrapper (fastify) {
...
+    preHandler: 'preHandler',
+    [false]: false
+  }
+
+  return hooks[hook] ?? next(new Error('@fastify/cookie: Invalid value provided for the hook-option. You can set the hook-option only to false, \'onRequest\' , \'preParsing\' , \'preValidation\' or \'preHandler\''))
+}
+
 function plugin (fastify, options, next) {
   const secret = options.secret
+  const hook = getHook(options.hook ?? 'onRequest', next)
```
[the remark this answers is not in the document]
Comment (Uzlopak) [reply plugin.js:80]: Yes you are right. Actually should not pass next into the getHook but getHook should return the correct value or an error and if it is an error, do the next(error) with return.

File: test/cookie.test.js (hunk 19 above)
Comment (vitaly-sazonov) [reply test/cookie.test.js:863]: what do you mean?

File: test/cookie.test.js [hunk 22]
```diff
@@ -847,3 +847,48 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+test('dont add default @fastify/cookie hook-function to fastify if hook-option to false  ', (t) => {
+  t.plan(6)
+  const fastify = Fastify()
+  fastify.register(plugin, { hook: false })
+
+  for (const hook of ['preValidation', 'preHandler']) {
+    fastify.addHook(hook, (req, reply, done) => {
+      t.equal(req.cookies, null)
+      done()
+    })
```
Comment (Eomm) [inline test/cookie.test.js:860]:
```suggestion
  for (const hook of ['preValidation', 'preHandler', 'preParsing']) {
    fastify.addHook(hook, async (req) => {
      t.equal(req.cookies, null)
    })
```

File: test/cookie.test.js (hunk 19 above)
Comment (Eomm) [reply test/cookie.test.js:863]: added a suggestion to clarify

File: plugin.js (hunk 17 above)
Comment (Uzlopak) [reply plugin.js:56]: But we know the arity of the hooks. Also spreading arguments is converting the iterable arguments into an array.

So we can also pass onReqHandlerWrapper the hook as second parameter and then do

```
function onReqHandlerWrapper (fastify, hook) {
  const donePosition = hook === 'preParsing' ? 3 : 2
  return function fastifyCookieHandler () {
    const fastifyReq = arguments[0]
    const done = arguments[donePosition]
    // [...]
```
no popping needed. 

Or like this:

```
function onReqHandlerWrapper (fastify, hook) {
  return hook === 'preParsing'
    ? function fastifyCookieHandler (fastifyReq, fastifyRes, payload, done) {
        fastifyReq.cookies = {} // New container per request. Issue #53
        const cookieHeader = fastifyReq.raw.headers.cookie
        if (cookieHeader) {
          fastifyReq.cookies = fastify.parseCookie(cookieHeader)
        }
        done()
      }
    : function fastifyCookieHandler (fastifyReq, fastifyRes, done) {
      fastifyReq.cookies = {} // New container per request. Issue #53
      const cookieHeader = fastifyReq.raw.headers.cookie
      if (cookieHeader) {
        fastifyReq.cookies = fastify.parseCookie(cookieHeader)
      }
      done()
    }
}
```

Comment (Uzlopak) [reply plugin.js:56]: Or

```
function onReqHandlerWrapper (fastify, hook) {
  function hookHandler (fastifyReq, done) {
    fastifyReq.cookies = {} // New container per request. Issue #53
    const cookieHeader = fastifyReq.raw.headers.cookie
    if (cookieHeader) {
      fastifyReq.cookies = fastify.parseCookie(cookieHeader)
    }
    done()
  }

  return hook === 'preParsing'
    ? function fastifyCookieHandler (fastifyReq, fastifyRes, payload, done) {
        hookHandler(fastifyReq, done)
      }
    : function fastifyCookieHandler (fastifyReq, fastifyRes, done) {
      hookHandler(fastifyReq, done)
    }
}

```

Comment (Uzlopak) [reply plugin.js:56]: I mean... creating an array and popping the last document etc. is probably not that performant :(

Comment (vitaly-sazonov) [reply plugin.js:56]: and if there are other hooks, we always know that NextFunction comes last.  

pop() has high overhead?

Comment (vitaly-sazonov) [reply plugin.js:56]: maybe so?

```
const fastifyReq = argv[0]
const done = argv[argv.length-1]
```

Comment (Uzlopak) [reply plugin.js:56]: creating an array to just pop the last element increases the overhead. Is it low? Is it high? Idk, but why should we not choose the implementation, which has the lowest overhead of them all. And one were i create less intermediary objects, is the better one (memory management)

And to be honest I prefer the second solution, because I know that it has the lowest overhead of all the impementations.

And yes we support currently only 4 hooks. And if we have to support one more, we can add it and check the arity. 

Tbh, when you created the Object for the getHook-check you wrote that it has faster access. I thought ok, if he wants that and it is faster no problem, despite the fact, that getHook will be called by a service only one time - at the startup. So the speed is not that relevant but faster startup is always desirable.

We talk here about a hook function. This will run on every request. So I expect a slow down with the proposed implementation.

Comment (Uzlopak) [reply plugin.js:56]: Didnt see that you wrote the question, while I was writing my comment.

the issue is that you always do `...argv` which is basically transforming the arguments iterable to an array. Then you do a argv.length, so you have to lookup the length and finally subtract by 1. Also you need  instantiate two variables.

And this will be done for every request, despite the fact, that argv.length will always be the same. 

I dont know what the others think about it ( @mcollina , @Eomm @climba03003 ) . Maybe wait for some feedback.

Comment (vitaly-sazonov) [reply plugin.js:56]: I agree, using a prepared handler will be faster than working with an array in every request. I'm for performance :) 

In the second solution, there is code repetition. It doesn’t look very good, but if you know for sure that it’s faster this way, then I’ll change the solution to this.

File: plugin.js [hunk 23]
```diff
@@ -50,19 +50,44 @@ function fastifyCookieClearCookie (reply, name, options) {
...
+    preHandler: 'preHandler',
+    [false]: false
   }
+
+  return hooks[hook]
 }
 
 function plugin (fastify, options, next) {
   const secret = options.secret
+  const hook = getHook(options.hook, next)
```
Comment (Uzlopak) [inline plugin.js:87]:
```suggestion
  const hook = getHook(options.hook)
```

File: plugin.js [hunk 24]
```diff
@@ -50,19 +50,44 @@ function fastifyCookieClearCookie (reply, name, options) {
...
+      fastifyReq.cookies = {} // New container per request. Issue #53
+      const cookieHeader = fastifyReq.raw.headers.cookie
+      if (cookieHeader) {
+        fastifyReq.cookies = fastify.parseCookie(cookieHeader)
+      }
+      done()
+    }
+}
+
+function getHook (hook = 'onRequest', next) {
```
Comment (Uzlopak) [inline plugin.js:73]:
```suggestion
function getHook (hook = 'onRequest') {
```

**Review by Uzlopak — CHANGES_REQUESTED** [review]

**Comment by jsumners** [comment]
> > But the middleware phase is executed before the onRequest phase. If you need to work with cookies in the middleware phase, then you will have to parse the cookies in the middleware itself, and after that the plugin will parse the cookies again before the route handler
> 
> What does that mean? Fastify does not have any such "phase" as Fastify does not use the "middleware" concept.

Do we have an answer to this question?

**Comment by vitaly-sazonov** [comment]
> > > But the middleware phase is executed before the onRequest phase. If you need to work with cookies in the middleware phase, then you will have to parse the cookies in the middleware itself, and after that the plugin will parse the cookies again before the route handler
> > 
> > 
> > What does that mean? Fastify does not have any such "phase" as Fastify does not use the "middleware" concept.
> 
> Do we have an answer to this question?

There is a middleware concept in the fastify documentation. But these concepts are also available in other fastify-based frameworks, for example, in NestJS. 

Tell me what you don't like?

File: types/plugin.d.ts [hunk 25]
```diff
@@ -125,8 +125,11 @@ declare namespace fastifyCookie {
     signed?: boolean;
   }
 
+  type HookType = 'onRequest' | 'preParsing' | 'preValidation' | 'preHandler'  | 'preSerialization';
+
   export interface FastifyCookieOptions {
     secret?: string | string[] | Signer;
+    hook?: HookType | false;
```
Comment (mcollina) [inline types/plugin.d.ts:132]: Would you mind adding a test for the new type?

**Comment by jsumners** [comment]
> There is a middleware concept in the fastify documentation.

Please link to this documentation.

> But these concepts are also available in other fastify-based frameworks, for example, in NestJS.

I know nothing of NestJS.

> Tell me what you don't like?

I am trying to understand how this is applicable to the Fastify ecosystem.

**Comment by vitaly-sazonov** [comment]
> 

https://www.fastify.io/docs/latest/Reference/Middleware/

The fact that middleware support was cut out of the kernel does not mean that middleware should not be used in applications. Here is the middleware connection package https://github.com/fastify/middie

File: types/plugin.d.ts (hunk 25 above)
Comment (vitaly-sazonov) [reply types/plugin.d.ts:132]: I can, but what is there to check?

**Comment by climba03003** [comment]
> The fact that middleware support was cut out of the kernel does not mean that middleware should not be used in applications. 

From my point of view, I really would like to deprecate both @fastify/express and @fastify/middie .

It only create problems, and users error instead of real beneficial code.

However, even in the core teams, we have both voice of depreciation and keep.

**Review by mcollina — APPROVED** [review]
LGTM. this change is worthwhile doing as most of our modules allow folks to customize when hooks are triggered (or if they are triggered at all).

File: types/plugin.d.ts (hunk 25 above)
Comment (Uzlopak) [reply types/plugin.d.ts:132]: Basically he want that you add following lines into plugin.test-d.ts

```typescript
const appWithHook: FastifyInstance = fastify();

appWithHook.register(cookie, { hook: false });
appWithHook.register(cookie, { hook: 'onRequest' });
appWithHook.register(cookie, { hook: 'preHandler' });
appWithHook.register(cookie, { hook: 'preParsing' });
appWithHook.register(cookie, { hook: 'preSerialization' });
appWithHook.register(cookie, { hook: 'preValidation' });
expectError(appWithHook.register(cookie, { hook: true }))
expectError(appWithHook.register(cookie, { hook: 'false' }))
```

**Comment by Uzlopak** [comment]
Please add following lines into plugin.test-d.ts. ALso import expectError from tsd (line 2 of plugin.test-d.ts)

```typescript
const appWithHook: FastifyInstance = fastify();

appWithHook.register(cookie, { hook: false });
appWithHook.register(cookie, { hook: 'onRequest' });
appWithHook.register(cookie, { hook: 'preHandler' });
appWithHook.register(cookie, { hook: 'preParsing' });
appWithHook.register(cookie, { hook: 'preSerialization' });
appWithHook.register(cookie, { hook: 'preValidation' });
expectError(appWithHook.register(cookie, { hook: true }))
expectError(appWithHook.register(cookie, { hook: 'false' }))
```

File: plugin.js [hunk 26]
```diff
@@ -50,19 +50,44 @@ function fastifyCookieClearCookie (reply, name, options) {
...
 
-function onReqHandlerWrapper (fastify) {
-  return function fastifyCookieOnReqHandler (fastifyReq, fastifyRes, done) {
-    fastifyReq.cookies = {} // New container per request. Issue #53
-    const cookieHeader = fastifyReq.raw.headers.cookie
-    if (cookieHeader) {
-      fastifyReq.cookies = fastify.parseCookie(cookieHeader)
+function onReqHandlerWrapper (fastify, hook) {
+  return hook === 'preParsing'
+    ? function fastifyCookieHandler (fastifyReq, fastifyRes, payload, done) {
```
Comment (Eomm) [inline plugin.js:55]: These two hook functions could be moved in the upper scope (you will save one closure)

**Review by Eomm — APPROVED** [review]
Last tiny suggestion

LGTM

File: types/plugin.d.ts (hunk 25 above)
Comment (vitaly-sazonov) [reply types/plugin.d.ts:132]: done

File: README.md [hunk 27]
```diff
@@ -33,7 +33,8 @@ const fastify = require('fastify')()
 
 fastify.register(require('@fastify/cookie'), {
   secret: "my-secret", // for cookies signature
-  parseOptions: {}     // options for parsing cookies
+  hook: 'onRequest', // disable cookie autoparsing or set autoparsing on any of the hooks: 'onRequest', 'preParsing', 'preHandler', 'preValidation'. default: onRequest
```
Comment (Uzlopak) [inline README.md:36]:
```suggestion
  hook: 'onRequest', // set to false to disable cookie autoparsing or set autoparsing on any of the following hooks: 'onRequest', 'preParsing', 'preHandler', 'preValidation'. default: 'onRequest'
```

File: test/cookie.test.js [hunk 28]
```diff
@@ -847,3 +847,99 @@ test('should not decorate fastify, request and reply if no secret was provided',
     message: 'req.unsignCookie is not a function'
   })
 })
+
+test('dont add default @fastify/cookie hook-function to fastify if hook-option to false  ', (t) => {
```
Comment (Uzlopak) [inline test/cookie.test.js:851]:
```suggestion
test('dont add auto cookie parsing to onRequest-hook if hook-option is set to false', (t) => {
```

File: test/cookie.test.js [hunk 29]
```diff
@@ -847,3 +847,99 @@ test('should not decorate fastify, request and reply if no secret was provided',
...
+  })
+})
+
+test('result in an error if hook-option is set to an invalid value', (t) => {
+  t.plan(1)
+  const fastify = Fastify()
+
+  t.rejects(
+    () => fastify.register(plugin, { hook: true }),
+    {},
```
Comment (Uzlopak) [inline test/cookie.test.js:885]:
```suggestion
```

**Review by Uzlopak — APPROVED** [review]
LGTM
